examples/chat.rs

```diff
@@ -0,0 +1,172 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A basic chat application with logs demonstrating libp2p and the gossipsub protocol
+//! combined with mDNS for the discovery of peers to gossip with.
+//!
+//! Using two terminal windows, start two instances, typing the following in each:
+//!
+//! ```sh
+//! cargo run
+//! ```
+//!
+//! Mutual mDNS discovery may take a few seconds. When each peer does discover the other
+//! it will print a message like:
+//!
+//! ```sh
+//! mDNS discovered a new peer: {peerId}
+//! ```
+//!
+//! Type a message and hit return: the message is sent and printed in the other terminal.
+//! Close with Ctrl-c.
+//!
+//! You can open more terminal windows and add more peers using the same line above.
+//!
+//! Once an additional peer is mDNS discovered it can participate in the conversation
+//! and all peers will receive messages sent from it.
+//!
+//! If a participant exits (Control-C or otherwise) the other peers will receive an mDNS expired
+//! event and remove the expired peer from the list of known peers.
+
+use futures::{prelude::*, select};
+use libp2p::{
+    core::muxing::StreamMuxerBox,
+    gossipsub, identity,
+    swarm::NetworkBehaviour,
+    swarm::{SwarmBuilder, SwarmEvent},
+    PeerId, Transport,
+};
+use rust_libp2p_nym::test_utils::create_nym_client;
+use rust_libp2p_nym::transport::NymTransport;
+use std::collections::hash_map::DefaultHasher;
+use std::error::Error;
+use std::hash::{Hash, Hasher};
+use std::time::Duration;
+use testcontainers::clients;
+use tokio::io;
+use tokio_util::codec;
+use tracing::info;
+use tracing_subscriber::EnvFilter;
+
+// We create a custom network behaviour that uses Gossipsub
+#[derive(NetworkBehaviour)]
+struct Behaviour {
+    gossipsub: gossipsub::Behaviour,
+}
+use libp2p::Multiaddr;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info")),
+        )
+        .init();
+    // Create a random PeerId
+    let id_keys = identity::Keypair::generate_ed25519();
+    let local_peer_id = PeerId::from(id_keys.public());
```

Comment:
Now that I see it, it seems we can reuse the same id_keys and peed_id used in the `gossipSub` and in the `NymTransport` and swarm. It still works though, but I'm not sure if this is sound.

Reply:
yeah, that should be fine, the peer ID isn't really used in the NymTransport, it's mostly to satisfy libp2p.